src/rules/optimize-image/optimize-image.ts

```diff
@@ -0,0 +1,124 @@
+        const endpoint: string = process.env.API_ENDPOINT; // eslint-disable-line no-process-env
+        const requestAsync = require('request-promise');
+
+        const notifyError = async (resource: string, error: any) => {
+            debug(`Error getting analyzing result for ${resource}.`, error);
+            await context.report(resource, null, `Couldn't get results for ${resource}. Error: ${error.message}`);
+        };
+
+        const validateTargetFetchEnd = (targetFetchEnd: ITargetFetchEnd) => {
+
+            const { resource }: { resource: string } = targetFetchEnd;
+
+            debug(`Validating rule.`);
+
+            if (!username) {
+                notifyError(resource, new Error('No username is provided for authentication.'));
+                failed = true;
+
+                return;
+            }
+
+            const options = {
+                auth: { user: username },
+                body: { url: resource },
+                json: true,
+                method: 'POST',
+                uri: endpoint
+            };
+
+            analyzePromise = requestAsync(options);
+            analyzePromise.catch(async (error) => {
+                failed = true;
+
+                await notifyError(resource, error);
+            });
+        };
+
+        const valiateScanEnd = async (fetchEnd: IScanEnd) => {
+            const { resource }: { resource: string } = fetchEnd;
+            const reportPromises: Array<Promise<void>> = [];
+
+            if (!analyzePromise || failed) {
+                return;
+            }
+
+            debug(`Waiting for the optimizing result of ${resource}.`);
+
+            const result: Result = await analyzePromise;
+
+            debug(`Received the optimizing result of ${resource}.`);
+
+            const { files }: { files: Array<File> } = result;
+            const unoptimized: Array<File> = files.filter((file) => {
+                if (!file) {
+                    return false;
+                }
+
+                const { bytes }: { bytes: Bytes } = file;
+
+                return bytes.output < bytes.input;
+            });
+
+            if (!unoptimized.length) {
+                debug(`All images are optimized.`);
+
+                return;
+            }
+
+            unoptimized.forEach((file) => {
+                const { bytes }: { bytes: Bytes } = file;
+                const sizeDiff: number = bytes.savings / 1024;
+                const percentageDiff: number = Math.round((bytes.savings / bytes.input) * 100);
+                const reportPromise: Promise<void> = context.report(file.url, null, `File "${file.name}" can be ${sizeDiff.toFixed(2)}kB (${percentageDiff}%) smaller.`);
+
+                reportPromises.push(reportPromise);
+            });
+
+            await Promise.all(reportPromises);
+        };
+
+        return {
+            'scan::end': valiateScanEnd,
+            'targetfetch::start': validateTargetFetchEnd
```

Comment:
`targetfetch::start` or `targetfetch::end`?
Do the api resolves the redirection? (from `http` to `https`). If so, let use `targetfetch::start`, if not, I think we need to use `targetfetch::end` to have the final url.